### PR TITLE
Add support for deerma.humidifier.jsq{s,5}

### LINKED ISSUE #1225: Mijia Smart Sterilization Humidifier - deerma.humidifier.jsq5
Before submitting a new request, use the search to see if there is an existing issue for the device.

**Device information:**

  - Name(s) of the device: Mijia Smart Sterilization Humidifier
  - Link: https://www.aliexpress.com/item/4000996247094.html?spm=a2g0o.search0304.0.0.a6ec51e14VlM8z&algo_pvid=0ddc29ce-d600-4c21-bc2a-f6599ed88419&algo_exp_id=0ddc29ce-d600-4c21-bc2a-f6599ed88419-0

Use `miiocli device --ip <ip address> --token <token>`.

Model: deerma.humidifier.jsq5
Hardware version: esp8266
Firmware version: 2.1.3

**Additional context**
probably a newer version of deerma.humidifier.jsq

"
Mijia Intelligent Sterilization Humidifier (EU version?) | deerma.humidifier.jsq | ZNJSQ01DEM | 4.5L, <=38dB, 300mL/h, 24W
Mijia Intelligent Sterilization Humidifier (EU version?) | deerma.humidifier.jsq5 | ZNJSQ01DEM | 4.5L, <=38dB, 300mL/h, 24W
"
Source: https://github.com/syssi/xiaomi_airpurifier

### PR DESCRIPTION
**Pull request by supar**
Support of the deerma.humidifier.jsqs

Model: deerma.humidifier.jsqs
Hardware version: esp8266
Firmware version: 2.1.3

Closes  #1225

### CODE REVIEW

**Comment by supar** [comment]
I need for the support of the "Mi Smart Humidifer S (deerma.humidifier.jsqs)". I'm not python developer at all so took example from airhumidifier_miot. I tested the code with my device.
I would be glad to get some feed back, or the help how to improve code if required.

**Comment by syssi** [comment]
Could you try to add some tests? Some bullet points to simplify the job:

1. Make a copy of https://github.com/rytilahti/python-miio/blob/master/miio/tests/test_airhumidifier_miot.py -> test_airhumidifier_jsqs.py
2. Update the `_INITIAL_STATE` to the state of your device
3. Drop useless tests. Add new ones / update the existing one to the methods of your device.

**Comment by manfred-mp** [comment]
Mi Smart Antibacterial Humidifier (deerma.humidifier.jsq5): https://home.miot-spec.com/spec?type=urn%3Amiot-spec-v2%3Adevice%3Ahumidifier%3A0000A00E%3Adeerma-jsq5%3A1

This is almost identical except SIID7;PIID 3 (overwet protect) is missing. Can this model be added at the same time, please?

**Comment by syssi** [comment]
Thanks a lot! This is good to go. If you want to make it perfect you could check this visualization:

https://app.codecov.io/gh/rytilahti/python-miio/compare/1193/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr%20comments&utm_term=Teemu%20R.#diff-bWlpby9haXJodW1pZGlmaWVyX2pzcXMucHk=

And expand the `miio/airhumidifier_jsqs.py` coverage report at the bottom. Every line indicated with a red bar is untested. You could add additional tests to test additional code paths to increase the test coverage of your new implementation.

**Review by syssi — APPROVED** [review]

File: miio/airhumidifier_jsqs.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+        if self.data["overwet_protect"] is not None:
+            return self.data["overwet_protect"]
+        return None
+
+
+class AirHumidifierJsqs(MiotDevice):
+    """Main class representing the air humidifier which uses MIoT protocol."""
+
+    mapping = _MAPPING
```
Comment (rytilahti) [inline miio/airhumidifier_jsqs.py:164]: Please add `_supported_models` here, it should be a list of model strings supported by this integration.

File: miio/airhumidifier_jsqs.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+    # Indicator Light
+
+    @property
+    def led_light(self) -> Optional[bool]:
+        """Return status of the LED."""
+
+        if self.data["led_light"] is not None:
+            return self.data["led_light"]
+        return None
```
Comment (rytilahti) [inline miio/airhumidifier_jsqs.py:132]: You can replace all similar structures with this:
```suggestion
        return self.data.get("led_light")
```

**Review by rytilahti — COMMENTED** [review]
Please create a new directory in under `miio.integration` and move all relevant code (incl. tests) under it. Perhaps `miio/integration/humidifier/deerma/`?

**Comment by fredronnv** [comment]
@supar I've got some time to work on this and would like to move this PR forward, as I'm testing it with deerma.humidifier.jsq5. @rytilahti I could submit a separate PR with the cleanups you've requested, if this is the best way forward?

**Comment by supar** [comment]
> @supar I've got some time to work on this and would like to move this PR forward, as I'm testing it with deerma.humidifier.jsq5. @rytilahti I could submit a separate PR with the cleanups you've requested, if this is the best way forward?

What is the difference between *.jsq**s** and jsq**5**

**Comment by fredronnv** [comment]
> > @supar I've got some time to work on this and would like to move this PR forward, as I'm testing it with deerma.humidifier.jsq5. @rytilahti I could submit a separate PR with the cleanups you've requested, if this is the best way forward?
> 
> What is the difference between *.jsq**s** and jsq**5**

As far as I can tell, the jsq**5** does not include overwet protection, otherwise everything from the code for jsq**s** functions as expected.

**Comment by supar** [comment]
@rytilahti i moved file to the separate directory as You suggested. But i'm not experienced how in one class keep support for two model which are different only in one property. Does it mean to make copy code of the properties, classes?

**Comment by rytilahti** [comment]
@supar you can either create a separate mapping and select the one to use for the query by checking the model (using `model` property), or better yet, have a single mapping and remove the unsupported property from the query just prior querying. The methods that set the state (actions or setters) for the unsupported property should raise an exception if the feature is not supported (based on that model information).

I'll take another review in the upcoming days but the code looks good on surface after the linting issues are solved. Btw, you can execute `pre-commit run -a` or `tox -e lint` on your local setup to perform the checks :-)

**Comment by tancou** [comment]
Hi,
I've tested with my `deerma.humidifier.jsq5` and it works great.
I've just the following warning 
```
WARNING:miio.device:Found an unsupported model 'deerma.humidifier.jsq5' for class 'AirHumidifierJsqs'. If this is working for you, please open an issue at https://github.com/rytilahti/python-miio/
```

I really hope this PR will be merged, and if @supar can take into consideration the `jsq5` variant, it could be awesome !

File: miio/__init__.py [hunk 3]
```diff
@@ -45,6 +45,7 @@
 from miio.integrations.fan.dmaker import Fan1C, FanMiot, FanP5, FanP9, FanP10, FanP11
 from miio.integrations.fan.leshow import FanLeshow
 from miio.integrations.fan.zhimi import Fan, FanZA5
+from miio.integrations.humidifier.deerma.airhumidifier_jsqs import AirHumidifierJsqs
```
Comment (rytilahti) [inline miio/__init__.py:48]:
```suggestion
from miio.integrations.humidifier.deerma import AirHumidifierJsqs
```

Comment (rytilahti) [reply miio/__init__.py:48]: Modify the `__init__.py` inside the deerma package to export this.

**Review by rytilahti — APPROVED** [review]
Please update the README.md and fix the export, and then this is good to go! :tada:

File: miio/integrations/humidifier/deerma/airhumidifier_jsqs.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    @property
+    def overwet_protect(self) -> Optional[bool]:
+        """Return True if overwet mode is active."""
+        return self.data.get("overwet_protect")
+
+
+class AirHumidifierJsqs(MiotDevice):
+    """Main class representing the air humidifier which uses MIoT protocol."""
+
+    _supported_models = ["deerma.humidifier.jsqs"]
```
Comment (rytilahti) [inline miio/integrations/humidifier/deerma/airhumidifier_jsqs.py:148]:
```suggestion
    _supported_models = ["deerma.humidifier.jsqs", "deerma.humidifier.jsq5"]
```
It is reportedly working with that device, so worth adding it already here.

Comment (tancou) [reply miio/integrations/humidifier/deerma/airhumidifier_jsqs.py:148]: This works for me and removes the warning I mentioned before.

**Comment by rytilahti** [comment]
This is good to go, thanks @supar!

**Pull request merged**